devito/symbolics/manipulation.py

```diff
@@ -319,12 +319,152 @@ def reuse_if_untouched(expr, args, evaluate=False):
+
+    mapper = transitive_closure(mapper)
+    input = [i.subs(mapper) for i in input]
+
     try:
-        bool(min(a, b))  # Can it be evaluated or simplified?
-        return min(a, b)
+        bool(Min(*input))  # Can it be evaluated or simplified?
+        simplified = Min(*input)
+        if not simplified.args:
+            return simplified
+        else:
+            input = list(simplified.args)
+            exp = MIN(input[0], input[1])
+            for i in input[2:]:
+                exp = MIN(exp, i)
+            return exp
     except TypeError:
-        return MIN(a, b)
+        exp = MIN(input[0], input[1])
+        for i in input[2:]:
+            exp = MIN(exp, i)
+        return exp
+
+
+def evalmax(input=None, assumptions=None):
+    """
+    Simplify evalmin if possible and return nested MIN/MAX expression.
+
+    Parameters
+    ----------
+    input : a list of the symbols to be simplified
+        The candidates to be evaluated. Defaults to None.
+    assumptions : A list of assumptions formed as relationals, assumed to be True
+        Defaults to None.
+
+    Examples
+    --------
+    'Assuming no values are know for a, b, c, d but we know that d<=a and c>=b we can'
+    'safely drop `a` and `c` from the cancidate list'
+
+    >>> from devito import Symbol
+    >>> a = Symbol('a')
+    >>> b = Symbol('b')
+    >>> c = Symbol('c')
+    >>> d = Symbol('d')
+
+    >>> evalmax([a, b, c, d], [[Le(d, a), Ge(c, b)]])
+    MAX(a, c)
+    """
+    if len(input) == 1:
+        return input[0]
+
+    mapper = {}
+    if assumptions:
+        for asm in assumptions:
+            if set(asm.args).issubset(input):
+                if (asm.__class__ in (Ge, Gt)):
+                    mapper.update({asm.args[1]: asm.args[0]})
+                elif (asm.__class__ in (Le, Lt)):
+                    mapper.update({asm.args[0]: asm.args[1]})
+
+    # Update mapper connections (graph vertices)
+    mapper = transitive_closure(mapper)
+    input = [i.subs(mapper) for i in input]
+
+    try:
+        bool(Max(*input))  # Can it be evaluated or simplified?
+        simplified = Max(*input)
+        if not simplified.args:
+            return simplified
+        else:
+            input = list(simplified.args)
+            exp = MAX(input[0], input[1])
+            for i in input[2:]:
+                exp = MAX(exp, i)
+            return exp
+    except TypeError:
+        exp = MAX(input[0], input[1])
+        for i in input[2:]:
+            exp = MAX(exp, i)
+        return exp
+
+
+def reachable_items(R, k):
+    try:
+        ans = R[k]
+        if ans != []:
+            ans = reachable_items(R, ans)
+        return ans
+    except:
+        return k
+
+
+def transitive_closure(R):
+    '''
+    Partially inherited from: https://www.buzzphp.com/posts/transitive-closure
+    Helps to collapse paths in a graph. In other words, helps to simplfiy a mapper's
+    keys and values when values also appears in keys.
+
+    Example
```

Comment:
example should be sympy-independent

Reply:
Not sure I understand this, do you mean with no Symbols?